packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/subscribe.ts

```diff
@@ -0,0 +1,294 @@
+  /**
+   * The SNS Topic to subscribe to.
+   */
+  readonly topic: sns.ITopic;
+
+  /**
+   * The props for creating a queue that will subscribe to the given topic.
+   * Please provide either the `queueProps` or the `queue`, but not both.
+   *
+   * @default none
+   */
+  readonly queueProps?: sqs.QueueProps;
+
+  /**
+   * The user-provided queue to subscribe to the given topic.
+   * Please provide either the `queueProps` or the `queue`, but not both.
+   *
+   * @default none
+   */
+  readonly queue?: sqs.IQueue;
+
+  /**
+   * The props for the Dead Letter Queue for the topic-queue.
+   *
+   * @default none
+   */
+  readonly deadLetterQueueProps?: DeadLetterQueueProps;
+}
+
+/**
+ * The settings for creating a Dead Letter Queue.
+ */
+export interface DeadLetterQueueProps {
+  /**
+   * The maximum number of times that a message can be received by consumers.
+   * When this value is exceeded for a message the message will be automatically sent to the Dead Letter Queue.
+   *
+   * @default 3
+   */
+  readonly maxReceiveCount?: number;
+
+  /**
+   * The number of seconds that Dead Letter Queue retains a message.
+   *
+   * @default Duration.days(14)
+   */
+  readonly retentionPeriod?: cdk.Duration;
+
+}
+
+/**
+ * A structure to record the subscriptions created by the extension.
+ */
+export interface Subscription {
+  /**
+   * The list of topics that the `subscriptionQueue` is subscribed to.
+   */
+  readonly topics: sns.ITopic[];
+
+  /**
+   * The subscription queue for the list of topics.
+   */
+  readonly subscriptionQueue: sqs.IQueue;
+}
+
+/**
+ * Settings for the hook which mutates the application container
+ * to add the subscription queue URLs to its environment.
+ */
+export interface ContainerMutatingProps {
+  /**
+   * The queue name and URLs to be added to the container environment.
+   */
+  readonly environment: { [key: string]: string };
+}
+
+/**
+ * This hook modifies the application container's environment to
+ * add the queue URLs for the subscription queues of the service.
+ */
+export class SubscribeMutatingHook extends ContainerMutatingHook {
+  private environment: { [key: string]: string };
+
+  constructor(props: ContainerMutatingProps) {
+    super();
+    this.environment = props.environment;
+  }
+
+  public mutateContainerDefinition(props: ecs.ContainerDefinitionOptions): ecs.ContainerDefinitionOptions {
+    return {
+      ...props,
+
+      environment: { ...(props.environment || {}), ...this.environment },
+    } as ecs.ContainerDefinitionOptions;
+  }
+}
+
+/**
+ * This extension creates a default SQS Queue, `eventsQueue`. It also supports creation of SQS Queue
+ * per SNS Topic and sets up the SNS Subscriptions accordingly.
+ *
+ * The created subscriptions can be accessed using `subscriptions` field and the
+ * default queue for this service can be accessed using the getter `<extension>.eventsQueue`.
+ */
+export class SubscribeExtension extends ServiceExtension {
```

Comment:
The name of the extension doesn't feel quite right. The naming methodology of CDK favors having nouns for the classes, and then verb methods on those classes. And I think in general we've tried to keep the same methodology for the extensions too. So it should be something like "service.add(noun)"

I'd suggest something like "Service.add(new IncomingQueue())" or something like that. You may have a better name idea as well, but think of nouns and adjectives for the name, rather than verbs.

Reply:
Updated the extension name to `EventsQueue()`, I felt we are already creating a default queue with name `eventsQueue` for the service so it would be more intuitive to name the extension the same. Let me know if that sounds good!